File: CPP/medium/longestPalindromicSubstring.cpp
#include <vector>
#include <string>
#include <iostream>

std::vector<int> getPalindrome(std::string s, int start, int end){
    while(start >= 0 && end < s.size() && s[start] == s[end]){
        start--;
        end++;
    }
    
    return {start + 1, end - 1};
}

std::string longestPalindrome(std::string s) {
    if(s.size() <= 1) return s;

    int start = 0, end = 0;

    for(int i = 0; i < s.size(); i++){
        // for palindrom with odd length
        std::vector<int> oddLen = getPalindrome(s, i, i);
        // for palindrom with an even length
        std::vector<int> evenLen = getPalindrome(s, i, i+1);
    
        if(oddLen[1] - oddLen[0] > (end - start)){
            start = oddLen[0];
            end = oddLen[1];
        }
        if(evenLen[1] - evenLen[0] > (end - start)){
            start = evenLen[0];
            end = evenLen[1];
        }
    }

    return s.substr(start, end - start + 1);
}

// s = a   Longest Palindromic Substring = a
// s = ac  Longest Palindromic Substring = a
// s = cbbd        Longest Palindromic Substring = bb
// s = babad       Longest Palindromic Substring = bab

int main(){
    std::vector<std::string> testSet = {"a", "ac", "cbbd","babad"};
    for(std::string test : testSet){
        std::cout << "s = " << test << "\tLongest Palindromic Substring = " << longestPalindrome(test)<< std::endl;
    }

    return 0;
}

File: Java/medium/longestPalindromicSubstring.java
import java.util.*;

class Solution {
    public static String longestPalindrome(String s) {
        if(s.length() <= 1) return s;
        int start = 0, end = 0;
        for(int i = 0; i < s.length(); i++){
            int[] odd = getPalindrome(s, i, i);
            int[] even = getPalindrome(s, i, i+1);

            if(odd[1] - odd[0] > end - start){
                end = odd[1];
                start = odd[0];
            }
            if(even[1] - even[0] > end - start){
                end = even[1];
                start = even[0];
            }
        }

        return s.substring(start, end + 1);
    }

    public static int[] getPalindrome(String s, int left, int right){
        int[] p = new int[2];

        while(left <= right && left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)){
            left--;
            right++;
        }
        p[0] = left + 1;
        p[1] = right - 1;

        return p;
    }

    // s = a   Longest Palindromic Substring = a
    // s = ac  Longest Palindromic Substring = a
    // s = cbbd        Longest Palindromic Substring = bb 
    // s = babad       Longest Palindromic Substring = bab

    public static void main(String[] args){
        String[] testSet = new String[]{"a", "ac", "cbbd","babad"};
        for(String test : testSet){
            System.out.println("s = " + test + "\tLongest Palindromic Substring = " + longestPalindrome(test));
        }
    
    }
}